### DISCUSSION TITLE
Function reorder tidak jalan pada ggplot dengan geom_col

### DISCUSSION THREAD

**Discussion by galuhchynintya** [teamalgoritma/community#885] [Q&A] [answered]
Halo tim algoritma,
saya mau bertanya, saya sedang membuat multivariate plot jenis bar plot, ini kenapa fungsi reorder tidak berjalan maksimal ya, ada 2 data yang terbalik, seperti pada gambar berikut.
![image](https://github.com/teamalgoritma/community/assets/165654927/2cebaf16-f1ef-4019-bf9e-6e200c791520)
Data retailer Kohl's dan Sports Direct terbalik urutannya. Apakah ada yang salah dengan syntax nya? Saya juga coba dengan menggunakan fungsi fct_reorder() seperti penjelasan pada diskusi https://github.com/teamalgoritma/community/discussions/421 , tapi hasilnya jadi lebih aneh.
Berikut syntax nya
#data untuk plot1
sales_region <- 
sales %>% 
  filter(season %in% "Winter",
         year %in% "2021") %>% 
  group_by(Retailer, Region) %>% 
  summarise(Total.Sales = sum(Total.Sales)) %>% 
  #arrange(-Total.Sales) %>% 
  mutate(label = glue("Retailer: {Retailer}
                      {Region} Sales: {comma(Total.Sales)}"))

# Pembuatan Visual Statis
plot1 <- 
ggplot(data = sales_region, 
       mapping = aes(x=Total.Sales, 
                     y=reorder(Retailer, Total.Sales),
                     text=label)) +
  geom_col(mapping = aes(fill=Region), position = "stack") +
  scale_fill_brewer(palette = "Set2")+
  scale_x_continuous(labels = comma)+
  labs(title = "Total Sales Based on Season",
       x = "Total Sales",
       y = "Retailer",
       fill="Region") +
  theme_classic()
# Mengubah Visual Statis menjadi Interaktif
ggplotly(plot1, tooltip = 'text')

Terima kasih

**Comment by Yusraf002** [comment]
Dear @galuhchynintya 

Terimakasih telah mengajukan pertanyaan melalui forum ini. 

Untuk kendala yang ibu alami, kami melihat untuk code yang diberikan sudah sesuai. disini ada beberapa opsi yang bisa kami sarankan:

1. bisa dicoba tampilkan terlebih dahulu visualisasi barplot biasa tanpa  informasi stack. kemudian periksa apakah hasil visual sudah terurut secara benar atau masih sama seperti kondisi diatas.
2. kiranya ibu berkenan untuk melampirkan file rmd dan dataset yang ibu gunakan untuk kami bantu solving lebih lanjut.

Bisa dicoba terlebih dahulu untuk opsi 1.  jika masih terkendala bisa kita diskusikan kembali.
Best Regards,
Team Algoritma

**Reply by Yusraf002** [reply]
```{r}
plot_opt.1 <- 
  ggplot(data = sales_region, 
         mapping = aes(
           x= Total.Sales,
           y= reorder(Retailer, Total.Sales), 
           text=label,
           
           # memindahkan warna stack ke canvas
           fill=Region)) +
                    
  geom_col(position = "stack") +
  scale_fill_brewer(palette = "Set2")+
  scale_x_continuous(labels = comma)+
  labs(title = "Total Sales Based on Season",
       x = "Total Sales",
       y = "Retailer",
       fill="Region") +
  theme_classic()
```

Plot opsi 1 diatas hanya untuk memeriksa apakah dengan memindahkan coloring barchart ke bagian canvas menjadikan reorder berfungsi dengan baik, apabila reorder memang tidak berhasil berikut ini adalah opsi lain untuk melakukan reorder bisa gunakan `fct_reorder()` untuk mengurutkan (default: ascending) dan `fct_rev()` (mengubah urutan menjadi descending) dari packages `forecats`.

```{r}
install.packages("forecats")
library(forecats)

plot_opt.2 <- 
  ggplot(data = sales_region, 
         mapping = aes(
           x = Total.Sales,
          
           # mengurutkan dengan packages forecats
           y = fct_rev(fct_reorder(Retailer,Total.Sales)), 
           text =label,
           
           # memindahkan warna stack ke canvas
           fill =Region)) +
                    
  geom_col(position = "stack") +
  scale_fill_brewer(palette = "Set2")+
  scale_x_continuous(labels = comma)+
  labs(title = "Total Sales Based on Season",
       x = "Total Sales",
       y = "Retailer",
       fill="Region") +
  theme_classic()
```

bisa di coba terlebih dahulu semoga berhasil

**Reply by galuhchynintya** [reply]
Ini hasil dari opsi 1 mas yusraf,
![image](https://github.com/teamalgoritma/community/assets/165654927/b7fd2b18-f985-4791-b9fd-d8a079d30dbf)
fungsi reorder masih belum berfungsi dengan baik

untuk opsi 2 dengan package forcats, hasilnya seperti ini
![image](https://github.com/teamalgoritma/community/assets/165654927/b6466743-63c9-4643-808f-4b5a89240e54)

**Reply by Yusraf002** [reply]
Selamat sore ibu Galuh,

Mengenai fungsi reorder ini boleh coba email di [mentor@algorit.ma](mailto:mentor@algorit.ma) dan melampirkan file zip project nya, agar nanti dapat kami bantu untuk periksa lebih lanjut.

**Reply by galuhchynintya** [reply]
ok mas yusraf, sudah di email

**Reply by Yusraf002** [reply]
Setelah kami periksa mengenai fungsi reorder disini, indikasi pertama fungsi reorder tidak mengurutkan dengan benar adalah karena pada data yang ibu miliki untuk setiap kategori Retailer berulang untuk setiap Region. 

Untuk mengatasi hal tersebut, kita bisa melakukan sorting secara manual dengan menggunakan fungsi 
`scale_y_discreate(limits = c(urutan retailer)` pada saat membuat plot statis. dengan catatan
 `**urutan retailer** dari **kiri ke kanan** akan ditampilkan pada bar chart menjadi **dari atas ke bawah**`

Sementara untuk mendapatkan urutan retailer, disini bisa kita lakukan group_by retailer dan summarise by total.sales terlebih dahulu.

```{r}
# menampilkan peringkat retailer by total sales
salreg2 %>% group_by(Retailer) %>% summarise(Total.Sales = sum(Total.Sales)) %>% arrange(Total.Sales) 
```

```{r}
# Pembuatan Visual Statis
plot1 <- 
ggplot(data = sales_region, 
       mapping = aes(x= Total.Sales, 
                     
                     # tidak menggunakan fungsi reorder
                     y= Retailer,
                     
                     # mewarnai stack by region
                     fill = Region,
                     
                     # memberi info tooltip
                     text=label)) +
  geom_col(position = "stack") +
  
  # mengurutkan berdasarkan custom order
  scale_y_discrete(limits = c("Walmart","Amazon","Kohl's","Sports Direct","Foot Locker","West Gear"))+
  
  scale_fill_brewer(palette = "Dark2")+
  scale_x_continuous(labels = unit_format(scale = 1e-6, suffix = " mio"),
                     breaks = seq(from = 0,
                                  to = 250e6,
                                  by = 50e6))+
  

  
  labs(title = "Total Sales Based on Season",
       x = "Total Sales",
       y = "Retailer") +
  theme_classic()+
  theme(plot.title = element_text(face = "bold", size = 14, hjust = 6))
# Mengubah Visual Statis menjadi Interaktif
ggplotly(plot1, tooltip = 'text')
```

bisa di coba terlebih dahulu semoga berhasil, apabila jawaban ini membantu boleh _**mark as answer**_ ya bu.

**Reply by galuhchynintya** [reply]
sudah sy coba mas yusraf, alhamdulillah berhasil jika memilih semua season dan semua tahun mas
tapi jadinya kalau saya hanya pilih season tertentu, tetap tidak urut karena urutan retailernya sudah ditentukan. Untuk file shiny nya saya kirimkan via email ya mas yusraf.

**Reply by Yusraf002** [reply, marked as answer]
Selamat sore Ibu Galuh,

Untuk melakukan filtering disini ibu bisa lakukan tahapan berikut:

1. preparasi data berisi filter musim dan tahun
2. membuat data agregasi untuk melihat urutan dan disimpan kedalam suatu objek
3. menambahkan scale_x/y_discreate pada pembuatan plot dan memasukkan limit dengan nama objek yang berisi urutan

dengan pengaplikasian sebagai berikut

**Data Preparation untuk Plot 1** 
```{r}
sales_region <- 
sales %>% 
  filter(season %in% "Winter",    # ganti musim disini
         year %in% "2021") %>%  # ganti tahun disini
  group_by(Retailer, Region) %>% 
  summarise(Total.Sales = sum(Total.Sales), .groups = 'drop') %>% 
  ungroup() %>% 
  #arrange(-Total.Sales) %>% 
  mutate(label = glue("Retailer: {Retailer}
                      {Region} Sales: {comma(Total.Sales)}"))
```

**Membuat agregasi urutan peringkat**
```{r}
urutan <- sales_region %>% aggregate(Total.Sales ~ Retailer, FUN = sum) %>% arrange(Total.Sales) %>%  mutate(Retailer = as.character(Retailer)) %>% select(Retailer) %>% as.list()

urutan
```

**Membuat visual dengan menambahkan scale_x/y_discreate**
```{r}
# Pembuatan Visual Statis
plot1 <- 
ggplot(data = sales_region, 
       mapping = aes(x=Total.Sales, 
                     y=reorder(Retailer, Total.Sales),
                     text=label)) +
  geom_col(mapping = aes(fill=Region), position = "stack") +
  scale_fill_brewer(palette = "Dark2")+

# menambahkan custom urutan
  scale_y_discrete(limits = urutan$Retailer)+
  scale_x_continuous(labels = unit_format(scale = 1e-6, suffix = " mio"),
                     breaks = seq(from = 0,
                                  to = 250e6,
                                  by = 50e6))+
  labs(title = "Total Sales Based on Season",
       x = "Total Sales",
       y = "Retailer",
       fill="Region") +
  theme_classic()+
  theme(plot.title = element_text(face = "bold", size = 14, hjust = 6))

# Mengubah Visual Statis menjadi Interaktif
ggplotly(plot1, tooltip = 'text')
```

Data preparasi dan agregasi urutan dijalankan bersamaan sebagai persiapan plot 1, dengan demikian setiap ibu memberikan filter musim dan tahun akan berhasil menyesuaikan urutan visualnya.

**Gambar 1: Winter, 2021**
![image](https://github.com/teamalgoritma/community/assets/127075678/91edc5f3-ebfd-4b65-a7fc-4e1d64c1cf60)

**Gambar 2: Summer, 2021**
![image](https://github.com/teamalgoritma/community/assets/127075678/c9428dda-8571-4c23-b6b8-e4620dc15071)


Demikian solusi yang dapat kami berikan, bisa di coba terlebih dahulu dan semoga berhasil. Apabila jawaban ini membantu boleh _**mark as answer**_ ya bu.

**Reply by galuhchynintya** [reply]
Alhamdulillah sudah bisa mas Yusraf, terima kasih banyak bantuannya.
